lib/libzpool/uzfs_io.c

```diff
@@ -512,3 +536,122 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
 	zv->zv_metavolblocksize = tgt_block_size;
 	return (0);
 }
+
+int
+uzfs_unmap_data(zvol_state_t *zv, uint64_t offset, uint64_t len,
+    blk_metadata_t *metadata, boolean_t is_rebuild)
+{
+	int error, ret = 0, i;
+	dmu_tx_t *tx;
+	itx_t *itx;
+	lr_truncate_t *lr;
+	metaobj_blk_offset_t metablk;
+	uint8_t *buf;
+	uint64_t moff, mend, mchunk, mwlen;
+	rl_t *rl;
+	uint32_t count = 0;
+	list_t *chunk_io = NULL;
+	uint64_t orig_offset = offset;
+	uint64_t end = len + offset;
+	uint64_t wrote = 0;
+	uint64_t blocksize = zv->zv_volblocksize;
+	uint64_t len_in_first_aligned_block = 0;
+
+	/*
+	 * If trying IO on fresh zvol before metadata granularity is set return
+	 * error.
+	 */
+
+	if (zv->zv_metavolblocksize == 0)
+		return (EINVAL);
+
+	ASSERT(metadata);
+	ASSERT3P(zv->zv_metavolblocksize, !=, 0);
+	if (!IS_P2ALIGNED(offset, zv->zv_metavolblocksize) ||
+	    !IS_P2ALIGNED(len, zv->zv_metavolblocksize) ||
+	    len == 0)
+		return (EINVAL);
+
+	if (offset + len > zv->zv_volsize || offset > zv->zv_volsize)
+		return (EINVAL);
+
+	rl = zfs_range_lock(&zv->zv_range_lock, offset, len, RL_WRITER);
+
+	if (is_rebuild) {
+		VERIFY(ZVOL_IS_DEGRADED(zv) && (ZVOL_IS_REBUILDING(zv) ||
+		    ZVOL_IS_REBUILDING_ERRORED(zv)));
+		count = uzfs_get_nonoverlapping_ondisk_blks(zv, offset,
+		    len, metadata, (void **)&chunk_io);
+		if (!count)
+			goto out;
+chunk_io:
+			GET_NEXT_CHUNK(chunk_io, offset, len, end);
+			wrote = offset - orig_offset;
+			CHECK_FIRST_ALIGNED_BLOCK(
+			    len_in_first_aligned_block, offset, len,
+			    blocksize);
+
+			zv->rebuild_info.rebuild_bytes += len;
+			count--;
+	}
+
+	error = dmu_free_long_range(zv->zv_objset, ZVOL_OBJ, offset, len);
+	if (error) {
+		/*
+		 * Ingore error as we are also writing free op in zil
+		 */
+		LOG_ERR("Failed to free range %lu+%lu err:%d\n",
+		    offset, len, error);
+	}
+
+	tx = dmu_tx_create(zv->zv_objset);
+	get_zv_metaobj_block_details(&metablk, zv, offset, len);
+	dmu_tx_hold_write(tx, ZVOL_META_OBJ, metablk.m_offset, metablk.m_len);
+
+	error = dmu_tx_assign(tx, TXG_WAIT);
+	if (error) {
+		dmu_tx_abort(tx);
+		ret = UZFS_IO_TX_ASSIGN_FAIL;
+		goto out;
+	}
+
+	if (!zil_replaying(zv->zv_zilog, tx)) {
+		itx = zil_itx_create(TX_TRUNCATE, sizeof (*lr));
```

Comment:
where are we storing io_num in the zil?

Reply:
Previously, we were not storing io_num in zil. In the new implementation, we are storing io_num in zil record.